### PR TITLE
3% speedup with LazyMap and MetadataStorage

### PR DESCRIPTION
**Pull request by JaroslavTulach**
### Pull Request Description

Shall optimize #8324 by making all deserialized `Map` implementations (both Java and Scala) lazy.

### Important Notes

I don't know how to convert Java `Map` to Scala immutable `Map` without copying - e.g. lazily. Scala experts please advice!

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.

### CODE REVIEW

File: engine/runtime-parser/src/test/java/org/enso/compiler/core/IrPersistanceTest.java [hunk 1]
```diff
@@ -61,6 +67,30 @@ public void scalaMap() throws Exception {
...
+  public void scalaImmutableMapIsLazy() throws Exception {
+    var s1 = new LazySeq("Hello");
+    var s2 = new LazySeq("World");
+    var in = (scala.collection.immutable.Map) scala.collection.immutable.Map$.MODULE$.empty()
+      .$plus(new Tuple2("Hello", s1))
+      .$plus(new Tuple2("World", s2));
+
+    LazySeq.forbidden = true;
+    var out = (scala.collection.immutable.Map)
+      serde(scala.collection.immutable.Map.class, in, 64);
```
Comment (JaroslavTulach) [inline engine/runtime-parser/src/test/java/org/enso/compiler/core/IrPersistanceTest.java:81]: The Java version below works. The Scala version fails with following exception:
```
sbt:runtime-parser> testOnly *IrPersistanceTest
 Test org.enso.compiler.core.IrPersistanceTest.scalaMapIsImmutable failed: java.lang.IllegalStateException: Cannot create LazySeq right now!, took 0.003 sec
     at org.enso.compiler.core.IrPersistanceTest$LazySeq.<init>(IrPersistanceTest.java:336)
     at org.enso.compiler.core.IrPersistanceTest$PersistLazySeq.readObject(IrPersistanceTest.java:391)
     at org.enso.compiler.core.IrPersistanceTest$PersistLazySeq.readObject(IrPersistanceTest.java:376)
     at org.enso.persist.Persistance.readWith(Persistance.java:144)
     at org.enso.persist.PerInputImpl.readInline(PerInputImpl.java:45)
     at org.enso.persist.PerBufferReference.readObject(PerBufferReference.java:27)
     at org.enso.persist.Persistance$Reference.get(Persistance.java:205)
     at org.enso.compiler.core.ir.IrLazyMap$En.getValue(IrLazyMap.java:59)
     at scala.collection.convert.JavaCollectionWrappers$JMapWrapperLike$$anon$5.next(JavaCollectionWrappers.scala:400)
     at scala.collection.convert.JavaCollectionWrappers$JMapWrapperLike$$anon$5.next(JavaCollectionWrappers.scala:397)
     at scala.collection.mutable.Growable.addAll(Growable.scala:62)
     at scala.collection.mutable.Growable.addAll$(Growable.scala:57)
     at scala.collection.immutable.MapBuilderImpl.addAll(Map.scala:690)
     at scala.collection.immutable.Map$.from(Map.scala:641)
     at scala.collection.immutable.Map.from(Map.scala)
     at org.enso.compiler.core.ir.IrPersistance$PersistScalaMap.readObject(IrPersistance.java:237)
     at org.enso.compiler.core.ir.IrPersistance$PersistScalaMap.readObject(IrPersistance.java:212)
     at org.enso.persist.Persistance.readWith(Persistance.java:144)
     at org.enso.persist.PerInputImpl.readInline(PerInputImpl.java:45)
     at org.enso.persist.PerBufferReference.readObject(PerBufferReference.java:27)
     at org.enso.persist.Persistance$Reference.get(Persistance.java:205)
     at org.enso.compiler.core.IrPersistanceTest.serde(IrPersistanceTest.java:315)
     at org.enso.compiler.core.IrPersistanceTest.scalaMapIsImmutable(IrPersistanceTest.java:81)
     at jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
     at java.lang.reflect.Method.invoke(Method.java:580)
```
What is the best way to convert `java.util.Map` to `scala.collections.immutable.Map`? @4e6, @hubertp?

Comment (4e6) [reply engine/runtime-parser/src/test/java/org/enso/compiler/core/IrPersistanceTest.java:81]: You can use [`CollectionConverters`](https://www.scala-lang.org/api/2.13.12/scala/jdk/javaapi/CollectionConverters$.html)

Comment (4e6) [reply engine/runtime-parser/src/test/java/org/enso/compiler/core/IrPersistanceTest.java:81]: But you cannot convert collections between Java and Scala without copying anyway

Comment (hubertp) [reply engine/runtime-parser/src/test/java/org/enso/compiler/core/IrPersistanceTest.java:81]: https://github.com/enso-org/enso/pull/8359/commits/f9c34a94dca334411f9565eca4cac9ef3fb5778d

File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/IrLazyImMap.scala [hunk 2]
```diff
@@ -1,55 +1,43 @@
...
-import scala.collection.immutable.MapOps
-import scala.collection.IterableFactory
-import scala.collection.immutable.Iterable
-import scala.collection.MapFactory
-import scala.annotation.unchecked.uncheckedVariance
-import scala.collection.mutable.Builder
+import scala.collection.Iterator
+
+final private class IrLazyImMap[K, V](
+  val underlying: java.util.Map[K, V]
```
Comment (JaroslavTulach) [inline engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/IrLazyImMap.scala:8]: Wow, thanks @hubertp for this code change! Looks like it is possible to have an _immutable Scala map_ that delegates to `java.util.Map`. Kudos to @thurka for coming up with this delegation idea. Thank you guys!

File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/IrLazyImMap.scala [hunk 3]
```diff
@@ -1,55 +1,43 @@
...
+}
 
-  override protected def newSpecificBuilder
-    : Builder[(K, V), IrLazyImMap[K, V]] @uncheckedVariance =
-    Map.newBuilder.mapResult((p: Map[K, V]) =>
-      new IrLazyImMap[K, V](p, defaultValue)
-    )
+object IrLazyImMap {
+  def apply[K, V](map: java.util.Map[K, V]): Map[K, V] =
+    new IrLazyImMap[K, V](map)
```
Comment (JaroslavTulach) [inline engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/IrLazyImMap.scala]: I guess I prefer to remove this and use `new IrLazyImMap(...)` at call sites. That's however a minor comment.

**Review by hubertp — APPROVED** [review]

File: engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/IrLazyImMap.scala (hunk 3 above)
Comment (JaroslavTulach) [reply engine/runtime-parser/src/main/scala/org/enso/compiler/core/ir/IrLazyImMap.scala]: Done in cac8910.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/TruffleCompilerContext.java [hunk 4]
```diff
@@ -374,12 +375,19 @@ public QualifiedName getName() {
...
-        var pass = meta.get(BindingAnalysis$.MODULE$);
-        return (BindingsMap) pass.get();
-      } else {
-        return bindings;
+        try {
+          var meta = module.getIr().passData();
+          var pass = meta.get(BindingAnalysis$.MODULE$);
+          emitIOException();
+          return (BindingsMap) pass.get();
+        } catch (IOException ex) {
```
Comment (JaroslavTulach) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/TruffleCompilerContext.java:388]: Lazy deserialization of `BindingsMap` may fail: https://github.com/enso-org/enso/actions/runs/6974085571/job/18979188994?pr=8359#step:10:1073 - don't propagate the `IOException` further.

File: engine/runtime-compiler/src/main/scala/org/enso/compiler/phase/ImportResolver.scala [hunk 5]
```diff
@@ -173,9 +173,21 @@ class ImportResolver(compiler: Compiler) {
...
+      if (b == null) {
+        compiler.context.updateModule(
+          mod,
+          { u =>
+            u.invalidateCache()
+            u.ir(null)
+            u.compilationStage(CompilationStage.INITIAL)
+          }
+        )
+        compiler.ensureParsed(mod)
```
Comment (JaroslavTulach) [inline engine/runtime-compiler/src/main/scala/org/enso/compiler/phase/ImportResolver.scala:186]: Reparse without caching when [lazy deserialization fails](https://github.com/enso-org/enso/actions/runs/6974085571/job/18979188994?pr=8359#step:10:1073).

**Comment by JaroslavTulach** [comment]
OK, we are green. Now we need to judge whether this change worths being integrated, @hubertp?

#### Cons

The `BindingsMap` may turn to be broken during its deserialization (because calling its `restoreFromSerialization` which may yield an `IOException` later when traversing the IR). Now, when we are doing _lazy deserialization_ the `IOException` signaling that `BindingsMap` cannot be created can be emitted at random places. When asking for `getMetadata(BindingAnalysis)`, or when calling `Module.getBindingsMap` it may fail.

#### Pros

We can gain some speed. The [base benchmark results](https://github.com/enso-org/enso/pull/8390#issuecomment-1826274914) are finally available. Let's schedule [benchmark run with this PR](https://github.com/enso-org/enso/actions/runs/6988557172) - on develop: `3983, 3985, 3968` ms - with this PR: `3855, 3858, 3872` ms. E.g. the **speedup is 100ms** - e.g. **3%**.

#### Decision

Today's a decision was made to integrate in spite of the "Cons".

**Review by jdunkerley — APPROVED** [review]

**Pull request merged**